client/src/main/java/io/pravega/client/netty/impl/FlowHandler.java

```diff
@@ -96,6 +97,33 @@ public void closeFlow(ClientConnection clientConnection) {
         flowIdReplyProcessorMap.remove(flow);
     }
 
+    /**
+     * Create a Batch size tracker on existing connection.
+     *
+     * @param id identifier.
+     */
+    public void createAppendBatchSizeTracker(final UUID id) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        if (idBatchSizeTrackerMap.containsKey(id)) {
+            log.info("Reusing Batch size tracker for id {}.", id);
+        } else {
+            log.info("Creating Batch size tracker for id {}.", id);
+            idBatchSizeTrackerMap.put(id, new AppendBatchSizeTrackerImpl());
+        }
+    }
+
+
+    /**
+     * get a Batch size tracker on existing connection.
+     *
+     * @param id identifier.
+     * @return Batch size Tracker object.
+     */
+    public AppendBatchSizeTracker getAppendBatchSizeTracker(final UUID id) {
```

Comment:
I don't think we need to tie this to the writerId. It can be based on the flowId. This will make initializing this map, and removing entries when they are no longer needed much simpler.
It will also have the added benefit that the non-append code paths aren't affected.

Reply:
Hi @tkaitchuck @shrids 
    The channelRead method of netty handler , calls the recordAck of append batch size tracker upon receiving the WireCommands.DataAppended . 
The DataAppended has the writerid;  so, if the appendbatchsizetracker per flow id; then we need to maintain the map between   "writer id -> flowid -> appendbatchsizetracker".

to do this,  writer id is required while creating the connection and flow id.
so, its better have the direct mapping of "writer id ->appendbatchsizetracker"

for append path, SegmentOutputStream   has to behave differently for state sycnhronizer and byte streamer.

with these changes,  out next changes will as as follows.

1. For state sycnhronizer - we can have SegmentOutputStream    with flowid and connection pooling.
2. for byte stream  - we can have SegmentOutputStream   with flowid disabled.


so,  i feel its better to keep the appendbatchsizetracker independent of flowid.

Reply:
To send the DataAooended command to the correct SegmentOutputStream.ReplyProcessor the FlowHandler is doing this:
```
        int flowId = disableFlow.get() ? FLOW_DISABLED : Flow.from(cmd.getRequestId()).getFlowId();
        final ReplyProcessor processor = flowIdReplyProcessorMap.get(flowId);
```
I'm not sure why we can't do the same thing to maintain a map to the appendBatchSizeTracker.

Reply:
@tkaitchuck ,
   yes , we can have this flowid -> appendsizeTracker ;
but the problem is this.
                     public void channelRead(ChannelHandlerContext ctx, Object msg) { 
                                  ....
                                  ....
                                    if (cmd instanceof WireCommands.DataAppended) {
                                                      final WireCommands.DataAppended dataAppended = 
                                                         (WireCommands.DataAppended) cmd;
                                   final AppendBatchSizeTracker batchSizeTracker = 
                                             getAppendBatchSizeTracker(dataAppended.getWriterId());
                                    if (batchSizeTracker != null) {
                                                       batchSizeTracker.recordAck(dataAppended.getEventNumber());
                                    }
        }  
                     }    


The  WireCommands.DataAppended   has the writerid , so we need mapping from writerid to appendbatchsizetracker.
so, we will end up mapping something like this
      writer id  -> flow id
and then
      flowid  -> appendbatchsize tracker.

however,  we can not avoid  writer id here.

Reply:
@kmgowda 
Every Segment writer has a unique request id which is created by using 
```
private final long requestId = Flow.create().asLong();
```
The same requestId is used when the SegmentStore replies to Append requests.
Hence we can use the same method used to fetch the correct `ReplyProcessor` to fetch the correct `AppendBatchTracker`

Reply:
@shrids 
But, requestID increments/changes in case of /pravega/client/src/main/java/io/pravega/client/segment/impl/ConditionalOutputStreamImpl.java [ConditionalOutputStreamImpl class]

<pre>
ConditionalOutputStreamImpl class
public boolean write(ByteBuffer data, long expectedOffset) throws SegmentSealedException {
     ....
     ....
 <b style="color:red" >  long requestId = client.getFlow().getNextSequenceNumber();   </b>
                            log.debug("Setting up append on segment: {}", segmentId);
                            SetupAppend setup = new SetupAppend(requestId, writerId,
                                                                segmentId.getScopedName(),
                                                                delegationToken);
                            val reply = client.sendRequest(requestId, setup);  
}
</pre>

so,  by calling **getNextSequenceNumber** and using it as **requestid** will create a  collision in the map of flowid to appendbatchsizeTracker.

Reply:
If you look at `Flow` I think that long comes from two ints, only one of which changes.

Reply:
Thanks Tom; yes, I will have map for flowid -> appendbatchsizeTracker..
changes are in progress.. I will update the PR ASAP.